File: static/js/diagnostics.js
/* Diagnostics view: component status panel, recent errors, live log tail.
   Auto-refreshes every 5 seconds while the view is mounted (CLAUDE.md). */

(function (global) {
  "use strict";

  var REFRESH_MS = 5000;
  var state = {
    timer: null,
    level: "",
    limit: 200,
  };

  function el(tag, cls, text) {
    var node = document.createElement(tag);
    if (cls) node.className = cls;
    if (text != null) node.textContent = text;
    return node;
  }

  function fmtTime(iso) {
    if (!iso) return "-";
    return String(iso).replace("T", " ");
  }

  function renderStatusPanel(container, data) {
    container.innerHTML = "";
    var rows = [
      { key: "database", name: "Database connection" },
      { key: "ai_extraction", name: "AI extraction (Gemini)" },
      { key: "anylist", name: "AnyList connection" },
    ];

    rows.forEach(function (r) {
      var d = data[r.key] || {};
      var row = el("div", "status-row");
      row.appendChild(el("span", "dot " + (d.state || "grey")));

      var body = el("div", "status-body");
      body.appendChild(el("div", "status-name", r.name));
      body.appendChild(el("div", "status-msg", d.message || ""));

      if (r.key === "ai_extraction") {
        body.appendChild(
          el(
            "div",
            "status-msg",
            "API enabled: " + (d.api_enabled ? "yes" : "no") + "  ·  fake mode: " + (d.fake_mode ? "yes" : "no")
          )
        );

        // Quota indicator — observed request count today per model (best-effort; Gemini's
        // free-tier caps aren't reliably documented — see CLAUDE.md > AI Provider Migration).
        var byModel = d.today_by_model || {};
        var quotaBits = Object.keys(byModel).map(function (m) {
          return m + ": " + byModel[m];
        });
        body.appendChild(
          el(
            "div",
            "status-msg",
            "Today: " + (quotaBits.length ? quotaBits.join("  ·  ") : "no calls") +
              "  ·  last success: " + fmtTime(d.last_success)
          )
        );

        if (d.dashboard_url) {
          var link = el("a", null, "open Google AI Studio quota dashboard →");
          link.href = d.dashboard_url;
          link.target = "_blank";
          link.rel = "noopener noreferrer";
          var linkWrap = el("div", "status-msg");
          linkWrap.appendChild(link);
          body.appendChild(linkWrap);
        }

        var q = d.queue || { depth: 0, items: [] };
        if (q.depth) {
          body.appendChild(
            el(
              "div",
              "status-msg",
              "Capture queue: " +
                q.depth +
                " item(s) waiting on quota — " +
                q.items
                  .map(function (it) {
                    return it.task + (it.attempt_count ? " (×" + it.attempt_count + ")" : "");
                  })
                  .join(", ")
            )
          );
        }

        var recent = d.recent_calls || [];
        if (recent.length) {
          var log = el("div", "ai-attempt-log");
          recent.forEach(function (c) {
            var line = el("div", "log-line");
            line.appendChild(el("span", "log-time", fmtTime(c.time)));
            line.appendChild(el("span", "log-level " + (c.outcome === "success" ? "INFO" : c.outcome === "quota" ? "WARNING" : "ERROR"), c.outcome));
            line.appendChild(
              el("span", "log-msg", c.task + " / " + c.model + (c.error_detail ? " — " + c.error_detail : ""))
            );
            log.appendChild(line);
          });
          body.appendChild(el("div", "status-msg", "Recent attempts:"));
          body.appendChild(log);
        }
      }
      if (r.key === "anylist") {
        body.appendChild(
          el(
            "div",
            "status-msg",
            "enabled: " + (d.enabled ? "yes" : "no") +
              "  ·  fake mode: " + (d.fake_mode ? "yes" : "no") +
              "  ·  list: " + (d.target_list || "?") +
              "  ·  creds: " + (d.credentials_configured ? d.secret_source || "yes" : "none")
          )
        );
        if (d.last_success)
          body.appendChild(el("div", "status-msg", "Last success: " + fmtTime(d.last_success)));
        if (d.last_push)
          body.appendChild(
            el(
              "div",
              "status-msg",
              "Last push: session " + d.last_push.session_id + " at " + fmtTime(d.last_push.pushed_at) +
                (d.last_push.confirmed ? " (confirmed)" : " — NOT confirmed")
            )
          );
        var checkBtn = el("button", null, "Check AnyList now");
        var checkOut = el("span", "status-msg");
        checkBtn.addEventListener("click", function () {
          checkBtn.disabled = true;
          checkOut.textContent = " checking…";
          api.diagnostics
            .anylistCheck()
            .then(function (res) {
              checkOut.textContent = " " + (res.ok ? "OK" : "FAILED") + " — " + res.detail;
              checkBtn.disabled = false;
            })
            .catch(function (err) {
              checkOut.textContent = " error: " + err.message;
              checkBtn.disabled = false;
            });
        });
        var checkWrap = el("div", "status-msg");
        checkWrap.appendChild(checkBtn);
        checkWrap.appendChild(checkOut);
        body.appendChild(checkWrap);
      }

      body.appendChild(el("div", "state-label", "state: " + (d.state || "grey")));
      row.appendChild(body);
      container.appendChild(row);
    });
  }

  function renderLogLines(container, entries) {
    container.innerHTML = "";
    if (!entries || entries.length === 0) {
      container.appendChild(el("div", "log-line", "no entries"));
      return;
    }
    entries.forEach(function (e) {
      var line = el("div", "log-line");
      line.appendChild(el("span", "log-time", fmtTime(e.time)));
      line.appendChild(el("span", "log-level " + e.level, e.level));
      line.appendChild(el("span", "log-msg", e.logger + ": " + e.message));
      container.appendChild(line);
    });
  }

  function refresh(root) {
    var statusPanel = root.querySelector("#diag-status");
    var errorsBox = root.querySelector("#diag-errors");
    var logList = root.querySelector("#diag-logs");
    var updated = root.querySelector("#diag-updated");

    api.diagnostics
      .status()
      .then(function (d) {
        if (statusPanel) renderStatusPanel(statusPanel, d);
      })
      .catch(function (err) {
        if (statusPanel) statusPanel.textContent = "Status unavailable: " + err.message;
      });

    api.diagnostics
      .recentErrors()
      .then(function (d) {
        if (errorsBox) renderLogLines(errorsBox, d.entries);
      })
      .catch(function () {});

    api.diagnostics
      .logs(state.limit, state.level || null)
      .then(function (d) {
        if (logList) renderLogLines(logList, d.entries);
        if (updated) updated.textContent = "updated " + new Date().toLocaleTimeString();
      })
      .catch(function (err) {
        if (logList) logList.textContent = "Logs unavailable: " + err.message;
      });
  }

  function mount(root) {
    root.innerHTML = "";

    // component status
    var statusCard = el("div", "card");
    statusCard.appendChild(el("h2", null, "Component status"));
    var statusPanel = el("div");
    statusPanel.id = "diag-status";
    statusPanel.textContent = "Loading...";
    statusCard.appendChild(statusPanel);
    root.appendChild(statusCard);

    // recent errors
    var errCard = el("div", "card");
    errCard.appendChild(el("h2", null, "Recent errors (last 10)"));
    var errBox = el("div", "log-list errors-box");
    errBox.id = "diag-errors";
    errCard.appendChild(errBox);
    root.appendChild(errCard);

    // log tail
    var logCard = el("div", "card");
    var head = el("h2", null, "Live log tail");
    logCard.appendChild(head);

    var controls = el("div", "log-controls");
    var levelSel = el("select");
    ["", "DEBUG", "INFO", "WARNING", "ERROR"].forEach(function (lv) {
      var opt = el("option", null, lv === "" ? "All levels" : lv);
      opt.value = lv;
      levelSel.appendChild(opt);
    });
    levelSel.value = state.level;
    levelSel.addEventListener("change", function () {
      state.level = levelSel.value;
      refresh(root);
    });
    controls.appendChild(levelSel);

    var refreshBtn = el("button", null, "Refresh now");
    refreshBtn.addEventListener("click", function () {
      refresh(root);
    });
    controls.appendChild(refreshBtn);

    var updated = el("span", "updated-note");
    updated.id = "diag-updated";
    controls.appendChild(updated);

    logCard.appendChild(controls);

    var logList = el("div", "log-list");
    logList.id = "diag-logs";
    logCard.appendChild(logList);
    root.appendChild(logCard);

    refresh(root);
    state.timer = global.setInterval(function () {
      refresh(root);
    }, REFRESH_MS);
  }

  function unmount() {
    if (state.timer) {
      global.clearInterval(state.timer);
      state.timer = null;
    }
  }

  global.DiagnosticsView = { mount: mount, unmount: unmount };
})(window);

File: app/routers/diagnostics.py
"""Diagnostics API — backs the /diagnostics page.

Provides:
  * GET  /api/v1/diagnostics/logs           live log tail (ring buffer)
  * GET  /api/v1/diagnostics/recent-errors  last 10 ERROR+ entries
  * GET  /api/v1/diagnostics/status         component status panel + AI quota / attempt log

AnyList indicator is a stub until Phase 5. Phase 3.9 M5 replaced the USD "spend tracker" +
its reset button with a Gemini daily-quota indicator + recent-attempt log (free tier, no
per-call cost).
"""

from __future__ import annotations

import json as _json
import logging

from fastapi import APIRouter, Depends, Query
from sqlalchemy import text
from sqlalchemy.orm import Session

from app.config import settings
from app.database import get_db
from app.log_config import get_log_entries
from app.models.history import ShoppingHistory
from app.models.queue import CaptureQueueItem
from app.services import ai_call_log
from app.services import anylist_client

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/diagnostics", tags=["diagnostics"])


@router.get("/logs")
def logs(
    limit: int = Query(200, ge=1, le=1000),
    level: str | None = Query(None, pattern="^(DEBUG|INFO|WARNING|ERROR|CRITICAL)$"),
    offset: int = Query(0, ge=0),
) -> dict:
    entries = get_log_entries(limit=limit + offset, level=level)
    if offset:
        entries = entries[offset:]
    return {"ok": True, "data": {"entries": entries[:limit], "count": len(entries[:limit])}}


@router.get("/recent-errors")
def recent_errors() -> dict:
    entries = get_log_entries(limit=1000, level="ERROR")[:10]
    return {"ok": True, "data": {"entries": entries}}


@router.get("/status")
def status(db: Session = Depends(get_db)) -> dict:
    """The component status panel: a `database`, `ai_extraction` (Gemini) and `anylist`
    block, each with a green/amber/red `state` + `message`. The AI block also carries the
    §0c enable/fake-mode flags, the daily quota indicator (`today_by_model`), the recent
    capture-attempt log, and the retry-queue summary. Every sub-query is wrapped so one
    failing block still returns the others."""
    # --- database -------------------------------------------------------
    db_ok = False
    try:
        db.execute(text("SELECT 1"))
        db_ok = True
    except Exception:  # noqa: BLE001
        logger.error("Diagnostics: database probe failed", exc_info=True)

    database = {
        "state": "green" if db_ok else "red",
        "message": "Connected" if db_ok else "Connection failed",
    }

    # --- AI extraction (Gemini) — Phase 3.9 M5 --------------------------------
    # Enable-switch / fake-mode fields are always populated (§0c). Quota = an observed
    # request count per model since local midnight; no dollar cost on Gemini's free tier.
    ai = {
        "state": "grey",
        "message": "Gemini API key not configured",
        "api_enabled": settings.ai_extraction_enabled,
        "fake_mode": settings.ai_extraction_fake_mode,
        "last_success": None,
        "today_by_model": {},
        "recent_calls": [],
        "queue": {"depth": 0, "items": []},
        "dashboard_url": "https://aistudio.google.com/app/apikey",
    }
    try:
        queued = (
            db.query(CaptureQueueItem)
            .order_by(CaptureQueueItem.queued_at.asc())
            .limit(20)
            .all()
        )
        ai["queue"] = {
            "depth": db.query(CaptureQueueItem).count(),
            "items": [
                {
                    "task": q.task,
                    "recipe_id": q.recipe_id,
                    "queued_at": q.queued_at.isoformat(timespec="seconds"),
                    "attempt_count": q.attempt_count,
                    "last_error": q.last_error,
                }
                for q in queued
            ],
        }
        ai["today_by_model"] = ai_call_log.today_counts_by_model(db)
        last_ok = ai_call_log.last_success_at(db)
        ai["last_success"] = str(last_ok) if last_ok is not None else None
        ai["recent_calls"] = [
            {
                "time": r.timestamp.isoformat(timespec="seconds"),
                "task": r.task,
                "model": r.model,
                "outcome": r.outcome,
                "error_detail": r.error_detail,
            }
            for r in ai_call_log.recent_calls(db, limit=15)
        ]

        if ai["fake_mode"]:
            ai["state"] = "amber"
            ai["message"] = "FAKE MODE — extraction returns canned fixtures, no real Gemini calls are made"
        elif not ai["api_enabled"]:
            ai["state"] = "grey"
            ai["message"] = "Disabled (AI_EXTRACTION_ENABLED=false in .env) — enable explicitly to use recipe capture"
        elif last_ok is not None:
            ai["state"] = "green"
            ai["message"] = "OK"
        elif settings.gemini_configured:
            ai["state"] = "amber"
            ai["message"] = "Key configured, no calls yet"
    except Exception:  # noqa: BLE001
        logger.error("Diagnostics: ai_call_log query failed", exc_info=True)

    # --- AnyList (Phase 5) -----------------------------------------
    # No live round-trip here (it would re-login on every auto-poll). Reports config state +
    # the last in-process success + the most recent push; POST /diagnostics/anylist-check
    # does an explicit auth round-trip on demand.
    anylist = {
        "enabled": settings.anylist_enabled,
        "fake_mode": settings.anylist_fake_mode,
        "target_list": settings.anylist_target_list_name,
        "credentials_configured": settings.anylist_configured,
        "secret_source": settings.anylist_secret_source,
        "last_success": None,
        "last_push": None,
        "state": "grey",
        "message": "",
    }
    try:
        anylist_last_ok = anylist_client.last_success_at()
        anylist["last_success"] = str(anylist_last_ok) if anylist_last_ok else None
        last_push_row = (
            db.query(ShoppingHistory).order_by(ShoppingHistory.pushed_at.desc()).first()
        )
        if last_push_row is not None:
            try:
                resp = _json.loads(last_push_row.anylist_response_json or "{}")
            except ValueError:
                resp = {}
            anylist["last_push"] = {
                "session_id": last_push_row.session_id,
                "pushed_at": str(last_push_row.pushed_at),
                "confirmed": not resp.get("discrepancies"),
            }
        if not settings.anylist_enabled and not settings.anylist_fake_mode:
            anylist["state"], anylist["message"] = "grey", "AnyList sync is switched off (ANYLIST_ENABLED=false)"
        elif settings.anylist_fake_mode:
            anylist["state"], anylist["message"] = "amber", "FAKE MODE — in-memory list, no real AnyList calls"
        elif not settings.anylist_configured:
            anylist["state"], anylist["message"] = "red", "Enabled but no credentials (see Security §2)"
        elif anylist["last_push"] and not anylist["last_push"]["confirmed"]:
            anylist["state"], anylist["message"] = "red", "Last push was not fully confirmed — check AnyList"
        elif anylist_last_ok:
            anylist["state"], anylist["message"] = "green", f"Last success {anylist_last_ok}"
        else:
            anylist["state"], anylist["message"] = "amber", "Enabled and configured; no call made yet"
    except Exception:  # noqa: BLE001
        logger.error("Diagnostics: anylist status build failed", exc_info=True)
        anylist["state"], anylist["message"] = "red", "status query failed"

    return {
        "ok": True,
        "data": {
            "database": database,
            "ai_extraction": ai,
            "anylist": anylist,
        },
    }


@router.post("/anylist-check")
def anylist_check() -> dict:
    """On-demand AnyList auth round-trip (fake or real) — the /diagnostics page's
    "Check AnyList now" button. Never raises; returns a timestamped status."""
    status = anylist_client.check_auth()
    return {
        "ok": True,
        "data": {"ok": status.ok, "detail": status.detail, "checked_at": str(status.checked_at)},
    }
